Feature static link lib download script

frayds opened the pull request:
add one function on downloading different files according to user's OS

Commit:
the static link lib download script was added

Commit:
the static link lib download script was added

Commit:
the static link lib download script was added

XadillaX committed:
chore(eslint): add eslint review (#1)

XadillaX commented:
Nits:

Rebase from the newest `first-demo` branch.

Some directories:

1. Linux: static library downloads to deps/lib/librocketmq.a
2. macOS: dynamic library downloads to deps/lib/librocketmq.dylib
3. Windows: dynamic and static libraries download to deps/lib/rocketmq-client-cpp.{dll,lib}

Commit message please follow [AngularJS's commit message convention](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines).

And use current ESLint to write code.

XadillaX commented:
/cc @duanpengfei

XadillaX committed:
feat: add setSessionCredentials

Commit:
q

Commit:
<chore>

<download static link lib according to uer's os>

<different static link lib is utilized when the program runs in different  os, so one script is  added>

Commit:
<chore>

<download static link lib according to uer's os>

<different static link lib is utilized when the program runs in different  os, so one script is  added>

XadillaX commented on script/staticLinkLibDownload.js:
```diff
@@ -0,0 +1,83 @@
+const ENVIRONMENT = process.env.npm_lifecycle_event;
```
`"use strict"`

XadillaX replied:
ps. filename use underscore: script/lib_download.js

Because macOS is dynamic library.

XadillaX commented on script/staticLinkLibDownload.js:
```diff
@@ -0,0 +1,83 @@
+const ENVIRONMENT = process.env.npm_lifecycle_event;
+
+if(ENVIRONMENT === "build") {
```
This script is certainly running under `postinstall`.

These `if` states is necessary
.

frayds replied:
```suggestion
if(ENVIRONMENT === "build") {
```
excuse me, i am still a little bit confused about what u said, could u please explain more, thanks.
as i understand, the lib must be downloaded before the `build` process and after `install` process, so i write the code in the `postInstall` process. why u said the `build` if states is necessary ?

XadillaX replied:
Sorry, these `if` states are **NOT** necessary. You can consider that this script is certainly running under `postInstall`.

XadillaX commented on script/staticLinkLibDownload.js:
```diff
@@ -0,0 +1,83 @@
+if(ENVIRONMENT === "postinstall") {
+    const os = require("os");
+    const osType = os.platform();
+    const httpclient = require("urllib").create();
```
```js
const urllib = require("urllib");

urllib.request(...);
```

XadillaX commented on script/staticLinkLibDownload.js:
```diff
@@ -0,0 +1,83 @@
+    const httpclient = require("urllib").create();
+    const fs = require("fs");
+
+    // 操作系统类型,返回值有'darwin', 'freebsd', 'linux', 'sunos' , 'win32'
```
Use English comment.

XadillaX commented on script/staticLinkLibDownload.js:
```diff
@@ -0,0 +1,83 @@
+    const fs = require("fs");
+
+    // 操作系统类型,返回值有'darwin', 'freebsd', 'linux', 'sunos' , 'win32'
+    if(osType === "win32") {
```
IMHO, we can modify it like:

```js
const pkg = require("../package");

const MIRROR_ROOT = "https://oss.../";
const OS_MAP = {
  macOS: "/something/like/librocketmq.dylib",
  linux: "..."
};
const version = pkg.cppSDKVersion;

const url = `MIRROR_ROOT/${version}${OS_MAP[osType]}`;
urllib.request(url, ...);
```

And then add a column named `cppSDKVersion` in package.json to specify the version.

If error occurred, we should `process.exit(4)`.

XadillaX replied:
btw, use English comment.

XadillaX reviewed (changes requested):
Some nits submitted.

XadillaX committed:
fix: message garbled under macOS (#4)

Commit:
feat: update the lib download script

Commit:
feat: update the lib download script

Commit:
feat: update the lib download script

XadillaX commented on script/lib_download.js:
```diff
@@ -0,0 +1,109 @@
+    windowsLIB: "rocketmq-client-cpp.lib"
+};
+
+if(ENVIRONMENT === "postinstall") {
```
we don't need `if` here.

XadillaX reviewed (changes requested):
```javascript
        httpclient.request(fileURL, {
            timeout: 1000000
        }, (err, data) => {
            if(err) {
                process.exit(4);
                throw err;
            }
            fs.writeFile(filePath, data, error => {
                if(error) {
                    process.exit(4);
                }
            });
        });
```
Code above is redundant.

```javascript
var Scarlet = require("scarlet-task");
var scarlet = new Scarlet(2);

function getUrlMap(callback) {
    const urlMap = [];
    switch(osType) {
    case "win32":
        urlMap.push(<dll>);
        urlMap.push(<lib>);
        process.nextTick(() => { callback(undefined, urlMap); });
        break;
    }

    case "darwin": {
        urlMap.push(<dylib>);
        process.nextTick(() => { callback(undefined, urlMap); });
        break;
    }

    case "linux": {
        osInfo.then(...);
        break;
    }

    default: process.nextTick(() => { callback(new Error('...')); });
    }
}

function download(task) {
    urllib.request(...);
}

getUrlMap(function(err, urlMap) {
    if(err) ...;

    for(const map of urlMap) scarlet.push(map, download);
    scarlet.afterFinish(urlMap.length, function() {
        // finished...
    }, false);
});
```

XadillaX commented:
btw, we can use `urllib`'s streaming feature to write a stream to a file stream.
